File: emulator/c/hw_she.c
// Sample Crypt primitives based on AES hash, CMAC and KDF for MARS
// Implements algorithms specified in AUTOSAR Secure Hardware Extensions (SHE)
// Author: Tom Brostrom

// AES implementation from:
// Tiny AES, https://github.com/kokke/tiny-AES-c

#include "../aes/aes.h" // Tiny AES

#include <string.h>   // for memset, memcpy
#include <stdio.h>
#include <stdbool.h>

#include "hw_she.h"

static void hexout(const char *msg, const void *buf, size_t len)
{
typeof(len) i;
    if (msg)
        printf("%s: ", msg);
    for (i=0; i<len; i++)
        printf("%02x", ((uint8_t *)buf)[i]);
    printf("\n");
}

// Left shift byte array by 1 bit
// Return the final carry bit
static uint8_t ls1(uint8_t *a, size_t n)
{
uint8_t C2, C = 0;

    while (n--)
        {
        C2 = a[n] >> 7;
        a[n] = (a[n] << 1) | C;
        C = C2;
        }
    return C;
}

// XOR Byte Arrays, x = x ^ y
static void xba(uint8_t * x, const uint8_t *y, size_t n)
{
    while (n--)
        x[n] ^= y[n];
}

// Simplified CMAC algorithm for a single block.
void SHE_cmac1(uint8_t *mac, const uint8_t *key, const uint8_t *blk)
{
struct AES_ctx ctx;

    AES_init_ctx(&ctx, key);
    memset(mac, 0, 16);
    AES_ECB_encrypt(&ctx, mac);
    if (ls1(mac,16))
        mac[15] ^= 0x87;
    xba(mac, blk, 16);
    AES_ECB_encrypt(&ctx, mac);
}

bool SHE_verify(const void *key, const void *dig, const void *sig)
{
uint8_t mac[16];
    SHE_cmac1(mac, key, dig);
    return memcmp(mac, sig, 16) == 0;
}

void SHE_kdf(void * key, const void * parent, char label, const void * ctx, uint16_t ctxlen)
{
she_hctx_t hctx;
    SHE_hash_init(&hctx);
    SHE_hash_update(&hctx, parent, 16);
    SHE_hash_update(&hctx, "\x01\x01", 2);
    SHE_hash_update(&hctx, &label, 1);
    SHE_hash_update(&hctx, ctx, ctxlen);
    SHE_hash_update(&hctx, "", 1);
    SHE_hash_fini(&hctx, key);
}


// Miyaguchi–Preneel (MP) compression
// https://en.wikipedia.org/wiki/One-way_compression_function#Miyaguchi–Preneel
// Apply MP round to each block in message M
// Any trailing incomplete block is ignored
static void mp_comp(const uint8_t * M, size_t n, uint8_t * H)
{
size_t j, b = n >> 4;  // number of full blocks to process, n / 16
struct AES_ctx ctx;
uint8_t EkM[16];

    while (b--)
        {
        AES_init_ctx(&ctx, H);
        memcpy(&EkM, M, sizeof(EkM));
        AES_ECB_encrypt(&ctx, EkM);
        for (j=0; j<16; j++)
            H[j] ^= EkM[j] ^ M[j];
        M += 16;
        }
}

// msg is padded in-place
// returns size of padded msg
static size_t pad(uint8_t * msg, size_t n, size_t total)
{
size_t i, r, z;
    if (!total)
        total = n;
    r = n & 0xf;                    // remainder bytes in last block of msg
    z = (r<=10 ? 10 : 26) - r;      // z = number of 0 bytes pad
    msg += n;
    *msg++ = 0x80;
    memset(msg, 0, z);
    msg += z;
    total <<= 3;                    // total *= 8, number of bits
    for (i=5; i-->0; )
        *msg++ = total >> (i<<3);   // append total bits, big endian in 5 bytes
    return n + 1 + z + 5;
}

void SHE_hash_init(she_hctx_t * hctx)
{
    memset(&hctx->H, 0, 16);
    hctx->total = 0;
    hctx->part_n = 0;
}

#define MIN(X, Y) (((X) < (Y)) ? (X) : (Y))

// Hash blocks from previous partial block (if any) and msg
// Bytes from trailing incomplete block in msg are copied to part_blk
void SHE_hash_update(she_hctx_t * hctx, const uint8_t * msg, size_t n)
{
size_t pn;  // number of bytes to append to partial block
    hctx->total += n;
    if (hctx->part_n)   // any bytes from partial block?
        {               // try to fill part_blk from msg
        pn = MIN(16-hctx->part_n, n);
        memcpy(&hctx->part_blk[hctx->part_n], msg, pn);
        hctx->part_n += pn;
        msg += pn;
        n -= pn;
        if (hctx->part_n == 16)
            {           // part_blk is full, compress it
            mp_comp(hctx->part_blk, 16, hctx->H);
            hctx->part_n = 0;
            }
        }
     if (n)     // additional bytes in msg
        {       // part_n should be 0 at this point
        mp_comp(msg, n, hctx->H);
        pn = n & 0xf;   // n % 16, bytes from trailing partial block
        msg += n - pn;  // move trailing pn bytes to part_blk
        hctx->part_n = pn;
        memcpy(&hctx->part_blk, msg, pn);
        }
}

// pad and do final compress, return digest
void SHE_hash_fini(she_hctx_t * hctx, void *dig)
{
    hctx->part_n = pad(hctx->part_blk, hctx->part_n, hctx->total);
    mp_comp(hctx->part_blk, hctx->part_n, hctx->H);
    memcpy(dig, hctx->H, sizeof(hctx->H));
}

void SHE_hash(void *out, const void * msg, size_t n)
{
she_hctx_t hctx;
    SHE_hash_init(&hctx);
    SHE_hash_update(&hctx, msg, n);
    SHE_hash_fini(&hctx, out);
}

// These tests are from the AUTOSAR SHE spec, 4.13 Examples and Test Vectors
// See https://www.autosar.org/fileadmin/user_upload/standards/foundation/20-11/AUTOSAR_TR_SecureHardwareExtensions.pdf
bool SHE_selftest(bool fullTest)
{
size_t i;
uint8_t key[16], out[16];
uint8_t key1[16] =  { 0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6,
                      0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c };
uint8_t msg1[16] =  { 0x6b, 0xc1, 0xbe, 0xe2, 0x2e, 0x40, 0x9f, 0x96,
                      0xe9, 0x3d, 0x7e, 0x11, 0x73, 0x93, 0x17, 0x2a };
uint8_t exp1[16] =  { 0x07, 0x0a, 0x16, 0xb4, 0x6b, 0x4d, 0x41, 0x44,
                      0xf7, 0x9b, 0xdd, 0x9d, 0xd0, 0x4a, 0x28, 0x7c };

uint8_t exp2[16] =  { 0x11, 0x8a, 0x46, 0x44, 0x7a, 0x77, 0x0d, 0x87,
                      0x82, 0x8a, 0x69, 0xc2, 0x22, 0xe2, 0xd1, 0x7e };

// TEST 1 for CMAC, from spec 4.13.2.3, example 1
    SHE_cmac1(out, key1, msg1);
    if (memcmp(out, exp1, 16) != 0)
        return false;

// TEST2 for KDF, from spec 4.13.2.5
    for (i=0; i<sizeof(key); i++)
        key[i] = i;
    SHE_kdf(out, key, 'S', "HE", 2);
    if (memcmp(out, exp2, 16) != 0)
        return false;

    return true;
}

File: emulator/python/hw_she.py
#!/usr/bin/env python3

# Sample Crypt primitives based on AES hash, CMAC and KDF for MARS
# Implements algorithms specified in AUTOSAR Secure Hardware Extensions (SHE)
# Author: Tom Brostrom, CPVI

# AES implementation from:
# https://github.com/ricmoo/pyaes/blob/master/pyaes/aes.py
import aes
import aes_mphash as hashmod

len_digest = hashmod.digest_size
len_sign = 16
len_skey = 16
alg_hash = 0x81  # TPM_ALG_? AES_MP

# convert int i to big endian array of n bytes
def int2bebar(i, n):
    return bytes([i>>(j<<3) & 0xff for j in reversed(range(n))])

# SHE only supports ECB (default) and CBC modes
# See tests below for examples
cipher_ecb = aes.AESModeOfOperationECB # (key).encrypt(data)
cipher_cbc = aes.AESModeOfOperationCBC # (key, iv).encrypt(data)

# Left shift bytearray by 1 bit. Return MSB.
def ls1(a):
    C = 0
    for i in reversed(range(len(a))):
        b = (a[i] << 1) | C
        a[i] = b & 0xff 
        C = b >> 8
    return C

# XOR Byte Arrays # compute x = x ^ y
def xba(x, y):
    assert len(x) == len(y)
    for i in range(len(x)):
        x[i] ^= y[i]

# Simplified CMAC algorithm for a single block.
def cmac1(K, blk):
    assert len(blk) == 16
    E = cipher_ecb(K).encrypt
    mtk = bytearray(E(bytes(16)))       # create the Tweak sub-key k0
    if ls1(mtk): mtk[15] ^= 0x87        # turn k0 into k1
    xba(mtk, blk)                       # create Mn' from k1 ^ blk
    return E(bytes(mtk))                # then encrypt Mn'

# Full CMAC algorithm.
# Verify with: https://artjomb.github.io/cryptojs-extension/
def cmac(K, msg):
    # print('key =', K.hex())
    E = cipher_ecb(K).encrypt
    mtk = bytearray(E(bytes(16)))       # create the Tweak sub-key k0
    if ls1(mtk): mtk[15] ^= 0x87        # turn k0 into k1

    f = len(msg) >> 4   # full blocks in msg
    r = len(msg) & 0xf  # remainder bytes in last block of msg
    # print('msg =', msg.hex(), " f =", f, ' r =', r)
    if r or not f:                      # partial block, or empty msg
        if ls1(mtk): mtk[15] ^= 0x87    # turn k1 into k2
        last = msg[-r:] + b'\x80' + bytes(15-r)
    else:
        f -= 1
        last = msg[-16:]

    # XOR the tweak sub-key (k1 or k2 in mtk) with last block
    xba(mtk, last)                      # Create Mn': mtk ^= last

    # process all but the last block
    V = bytearray(16)
    for i in range(f):
        xba(V, msg[i<<4:(i+1)<<4])      # V ^= Mi
        V = bytearray(E(bytes(V)))      # V = E(V)

    # process the last tweaked block Mn' in mtk
    xba(V, mtk)                         # V ^= Mn'
    V = E(bytes(V))                     # V = E(V)
    return V

def CryptHash(msg):
    return hashmod.new(msg).digest()

CryptSign = cmac1

def CryptVerify(key, dig, sig):
    return sig == CryptSign(key, dig)

# Key Derivation Function
def shekdf(K, C):
    return CryptHash(K + b'\x01\x01' + C + b'\x00')

def CryptSkdf(key, x, y):
    return shekdf(key, x + y)

CryptAkdf = None

def CryptSelfTest(fullTest):
    return True # TODO: write some real tests

if __name__ == "__main__":

    # ECB TESTS
    key = bytes.fromhex('000102030405060708090a0b0c0d0e0f')
    c = cipher_ecb(key)
    pt  = bytes.fromhex('00112233445566778899aabbccddeeff')
    exp = bytes.fromhex('69c4e0d86a7b0430d8cdb78070b4c55a')
    out = c.encrypt(pt)
    print('ECB1 TEST:', 'pass' if out == exp else 'FAIL')
    out = c.decrypt(out)
    print('ECB2 TEST:', 'pass' if out == pt else 'FAIL')

    # CBC TESTS

    key = bytes.fromhex('2b7e151628aed2a6abf7158809cf4f3c')
    iv  = bytes.fromhex('000102030405060708090a0b0c0d0e0f')
    e = cipher_cbc(key, iv).encrypt

    pt  = bytes.fromhex('6bc1bee22e409f96e93d7e117393172a')
    exp = bytes.fromhex('7649abac8119b246cee98e9b12e9197d')
    out = e(pt)
    print('CBC1 TEST:', 'pass' if out == exp else 'FAIL')

    pt  = bytes.fromhex('ae2d8a571e03ac9c9eb76fac45af8e51')
    exp = bytes.fromhex('5086cb9b507219ee95db113a917678b2')
    out = e(pt)
    print('CBC2 TEST:', 'pass' if out == exp else 'FAIL')
    
    # HASH TESTS

    msg = bytes.fromhex(
        '6bc1bee22e409f96e93d7e117393172aae2d8a571e03ac9c9eb76fac45af8e51')
    exp = bytes.fromhex('c7277a0dc1fb853b5f4d9cbd26be40c6')
    out = CryptHash(msg)
    print('HASH TEST:', 'pass' if out == exp else 'FAIL')

    hobj = hashmod.new()
    hobj.update(msg[0:4])
    hobj.update(msg[4:21])
    hobj.update(msg[21:32])
    out = hobj.digest()
    print('SEQ TEST:', 'pass' if out == exp else 'FAIL')

    # CMAC TESTS

    K   = bytes.fromhex('2b7e151628aed2a6abf7158809cf4f3c')
    msg = bytes.fromhex('6bc1bee22e409f96e93d7e117393172a')
    exp = bytes.fromhex('070a16b46b4d4144f79bdd9dd04a287c')
    out = cmac(K, msg)
    print('CMAC TEST:', 'pass' if out == exp else 'FAIL')
 
    out = cmac1(K, msg)
    print('CMAC1 TEST:', 'pass' if out == exp else 'FAIL')

    # SIGN TESTS
    h = hashmod.new()
    h.update(b'check 1 2 3')
    sig = CryptSign(K, h.digest())

    h = hashmod.new()
    h.update(b'check 1 2 3')
    print('SIGN TEST:', 'pass' if CryptVerify(K, h.digest(), sig) else 'FAIL')

    # KDF TEST

    K   = bytes.fromhex('000102030405060708090a0b0c0d0e0f')
    exp = bytes.fromhex('118a46447a770d87828a69c222e2d17e')
    out = CryptSkdf(K, b'SHE', b'')
    print('KDF TEST:', 'pass' if out == exp else 'FAIL')
